liquidctl/driver/coolit.py

```diff
@@ -0,0 +1,313 @@
+"""liquidctl driver for Corsair Platinum and PRO XT coolers.
+
+Supported devices
+-----------------
+
+ - Corsair H110i GT
+
+Supported features
+------------------
+
+ - general monitoring
+ - pump speed control
+ - fan speed control
+
+Copyright (C) 2020–2020  Jonas Malaco
+Copyright (C) 2020–2020  each contribution's author
+
+SPDX-License-Identifier: GPL-3.0-or-later
+"""
+
+import itertools
+import logging
+
+from enum import Enum, unique
+
+from liquidctl.driver.usb import UsbHidDriver
+from liquidctl.keyval import RuntimeStorage
+from liquidctl.pmbus import compute_pec
+from liquidctl.util import clamp, fraction_of_byte, u16le_from, normalize_profile
+
+
+LOGGER = logging.getLogger(__name__)
+
+_REPORT_LENGTH = 64
+_MAX_FAN_RPM = 2250
+_PUMP_INDEX = 0x02
+
+_COMMAND_FIRMWARE_ID = 0x01
+_COMMAND_TEMP_READ = 0x0E
+_COMMAND_FAN_SELECT = 0x10
+_COMMAND_FAN_MODE = 0x12
+_COMMAND_FAN_FIXED_PWM = 0x13
+_COMMAND_FAN_FIXED_RPM = 0x14
+_COMMAND_FAN_READ_RPM = 0x16
+_COMMAND_FAN_RPM_TABLE = 0x19
+_COMMAND_FAN_TEMP_TABLE = 0x1A
+
+_OP_CODE_WRITE_ONE_BYTE = 0x06
+_OP_CODE_READ_ONE_BYTE = 0x07
+_OP_CODE_WRITE_TWO_BYTES = 0x08
+_OP_CODE_READ_TWO_BYTES = 0x09
+_OP_CODE_WRITE_THREE_BYTES = 0x0A
+_OP_CODE_READ_THREE_BYTES = 0x0B
+
+_PROFILE_LENGTH = 5
+_CRITICAL_TEMPERATURE = 60
+
+_PUMP_DEFAULT_QUIET = [0x2e, 0x09]
+_PUMP_DEFAULT_EXTREME = [0x86, 0x0b]
+
+@unique
+class _FanMode(Enum):
+    FIXED_DUTY = 0x02
+    FIXED_RPM = 0x04
+    CUSTOM_PROFILE = 0x0E
+
+    @classmethod
+    def _missing_(cls, value):
+        LOGGER.debug("falling back to FIXED_DUTY for _FanMode(%s)", value)
+        return _FanMode.FIXED_DUTY
+
+
+@unique
+class _PumpMode(Enum):
+    QUIET = 0x08
+    EXTREME = 0x0C
+
+    @classmethod
+    def _missing_(cls, value):
+        LOGGER.debug("falling back to QUIET for _PumpMode(%s)", value)
+        return _PumpMode.QUIET
+
+
+def _sequence(storage):
```

Comment:
You mentioned that perfectly increasing and cycling sequence numbers might not be required, based on how the protocol looks like.

If this is right and the device responds correctly, It would be cheaper to use a simple cycling iterator initialized with a random value, instead of read/writing to `self._data`.

Or maybe even sending a constant (only randomized at `connect` time) is fine if the device doesn't complain and we end-up not checking the ACKs.

Reply:
From the looks of it I doubt this was supposed to be a global sequence, but rather a per-command sequence. I hacked it a bit to always start with `1`, and it still works.